rfcs/v3/authorization-rules.md

```diff
@@ -0,0 +1,568 @@
+  - Conditional activation of rules. Conditions support a basic set of operators (`equals`, `isNull`, and logical operators at least).
+  - Value expressions can either be literals or session variables only.
+  - Support only field permissions / model select predicates / argument presets. No input validation.
+  - Composability of rules.
+- New features (in no particular order):
+  - Reusing relationship permissions
+  - Arguments input validation
+  - Calling external systems
+  - Accessing sub-fields in value expressions
+  - Advanced boolean operators
+
+## FAQs
+
+1. **Will existing metadata continue to be work?**  
+    
+    Yes, existing metadata (using the `v1` version of permissions objects) will continue to work and trivially maps to the new system where the condition is an equality check on the `x-hasura-role` session variable.
+
+2. **Does this subsume the inherited roles feature in v2?**
+    
+    Yes, this is the intended replacement for inherited roles in v2. So, instead of defining multiple base roles and an inherited you would simply define multiple rules which can activated by arbitrary conditions.
+
+    Note, however, that cell-level authorization will not be supported by this system, since you define type (column) permissions and model (row) permissions independently.
+
+3. **Are any changes required to the NDC protocol?**
+   
+   No.
+
+4. **Are any changes required to the user's authentication system?**
+
+   No changes are required to existing authentication systems. However, to fully utilize the power of authorization rules you would want to inject more granular claims (instead of simply roles) during authentication.
+
+## Before / After comparison
+
+Imagine the following data models for an internal API targeted at a company's own developers and support agents.
+
+- User: id, name, email, is_gov
+- UserActivityRecord: id, user_id, details, is_hidden
+- SupportTickets: id, assigned_agent_id
+
+And these are the potential authorization considerations:
+
+- Two personas: Developers and Support Agents
+  - Developers can generally view all user's information
+  - Support agents can only view data for users whose tickets they have been assigned
+- Some user information is protected
+  - PII fields (name, email) unless the API consumer has explicit access
+  - Government user data is restricted unless the API consumer has explicit access
+
+### Before
+
+You would need to define the following roles and provision them appropriately in the authentication system.
+
+- developer_base
+- developer_with_pii_access
+- developer_with_gov_access
+- developer_with_pii_access_and_gov_access
+- support_agent_base
+- support_agent_with_pii_access
+- support_agent_with_gov_access
+- support_agent_with_pii_access_and_gov_access
+
+And the permissions would be very hard to manage. Eg:
+
+```yaml
+kind: TypePermission
+version: v1
+definition:
+  typeName: User
+  roles:
```

Comment:
I know this isn't the root concern of this RFC, but surely there are other ways to make such a permission less arduous to define, for instance?

```yaml
  permissions:
    - roles:
      - developer_base
      - developer_with_gov_access
      - support_agent_base
      - support_agent_with_gov_access
      output:
        allowedFields: [id]
    - roles:
      - developer_with_pii_access
      - developer_with_pii_access_and_gov_access
      - support_agent_with_pii_access
      - support_agent_with_pii_access_and_gov_access
      output:
        allowedFields: [id, name, email]
```

Reply:
Right, for type permissions itself, it's not as big a deal, but you would still need to model 8 different roles in your authn/authz system. And now imagine, another attribute like 'hippa_access' gets added, now I would have 16 roles to manage.

So, fundamentally the issue is that it's not always feasible to enumerate all possible authz states that can exist in my system since that can grow exponentially with each independent attribute in my system.